src/Compilers/CSharp/Portable/Symbols/FileIdentifier.cs

```diff
@@ -3,39 +3,93 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
 using System.Text;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis.CSharp;
 
-internal readonly struct FileIdentifier
+internal sealed class FileIdentifier
 {
+    private record FileIdentifierData(string? EncoderFallbackErrorMessage, string DisplayFilePath, ImmutableArray<byte> FilePathChecksumOpt);
+
     private static readonly Encoding s_encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
 
-    public string? EncoderFallbackErrorMessage { get; init; }
-    public ImmutableArray<byte> FilePathChecksumOpt { get; init; }
-    public string DisplayFilePath { get; init; }
+    private readonly string _filePath;
```

Comment:
>_filePath [](http://example.com/codeflow?start=28&length=9)

Are we now holding this string longer than necessary?

Reply:
Yeah, I saw that too, and it seemed like a decent tradeoff to make since I rarely see the FileIdentifierData actually realized and it simplified the concern about needing to have ordered reads/writes of various members. If you think it's not a good tradeoff, I can switch it back and add (what I think are) the appropriate memory barriers.

Reply:
> Yeah, I saw that too, and it seemed like a decent tradeoff to make since I rarely see the FileIdentifierData actually realized and it simplified the concern about needing to have ordered reads/writes of various members. 

Fair enough